src/models/blockchain/transaction/validity/single_proof.rs

```diff
@@ -794,49 +794,98 @@ mod test {
         }
     }
 
+    fn positive_prop(witness: SingleProofWitness) {
+        let claim = witness.claim();
+        let public_input = PublicInput::new(claim.input);
+        let rust_result = SingleProof.run_rust(&public_input, witness.nondeterminism());
+        let tasm_result = SingleProof.run_tasm(&public_input, witness.nondeterminism());
+        assert_eq!(rust_result.unwrap(), tasm_result.unwrap());
+    }
+
     mod proof_collection_tests {
+        use tasm_lib::hashing::merkle_verify::MERKLE_AUTHENTICATION_ROOT_MISMATCH_ERROR;
+
         use super::*;
 
         #[tokio::test]
-        async fn can_verify_via_valid_proof_collection() {
+        async fn disallow_set_merge_bit() {
             let mut test_runner = TestRunner::deterministic();
-            let primitive_witness = PrimitiveWitness::arbitrary_with_size_numbers(Some(2), 2, 2)
-                .new_tree(&mut test_runner)
-                .unwrap()
-                .current();
-            let txk_mast_hash = primitive_witness.kernel.mast_hash();
+            let good_primitive_witness =
+                PrimitiveWitness::arbitrary_with_size_numbers(Some(6), 6, 7, false)
+                    .new_tree(&mut test_runner)
+                    .unwrap()
+                    .current();
 
-            let proof_collection = ProofCollection::produce(
-                &primitive_witness,
+            let good_proof_collection = ProofCollection::produce(
+                &good_primitive_witness,
                 &TritonVmJobQueue::dummy(),
                 TritonVmJobPriority::default().into(),
             )
             .await
             .unwrap();
-            assert!(proof_collection.verify(txk_mast_hash));
 
-            let single_proof_witness = SingleProofWitness::from_collection(proof_collection);
-            let txk_mast_hash_as_input_as_public_input =
-                PublicInput::new(txk_mast_hash.reversed().values().encode());
+            let good_witness = SingleProofWitness::from_collection(good_proof_collection);
+            positive_prop(good_witness);
 
-            let nondeterminism = single_proof_witness.nondeterminism();
+            // Setting the `merge_bit` must make program crash, as this bit may
+            // only be set to true through the execution of the merge branch.
+            let bad_primitive_witness =
+                PrimitiveWitness::arbitrary_with_size_numbers(Some(6), 6, 7, true)
```

Comment:
As an alternative to passing everything through the function interface, consider sampling an arbitrary object and then modifying it. The `merge_bit` is rather free of internal consistency dependencies, so I don't think it will be a problem.

Reply:
> As an alternative to passing everything through the function interface, consider sampling an arbitrary object and then modifying it. The `merge_bit` is rather free of internal consistency dependencies, so I don't think it will be a problem.

This was my 1st attempt. But `PrimitiveWitness` witness data in the form of authentication paths which will be invalid if I flip the `merge_bit` after the witness has been produced.